File: internal/compose/state_test.go
package compose

import "testing"

const oneImage = `{"loading":false,"streaming":false,
"imageURLs":["https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs&sig=x"],
"alts":["Generated image: Teal Mountain"]}`

func TestParseState(t *testing.T) {
	s, err := ParseState([]byte(oneImage))
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if s.Loading || s.Streaming {
		t.Fatalf("flags wrong: %+v", s)
	}
	if len(s.ImageURLs) != 1 {
		t.Fatalf("images = %v", s.ImageURLs)
	}
}

func TestDistinctImageIDs(t *testing.T) {
	// The spike saw the same generated image rendered by three <img> tags.
	s := PageState{ImageURLs: []string{
		"https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs",
		"https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs",
		"https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs",
		"https://chatgpt.com/cdn/assets/favicon-x.svg",
	}}
	got := s.DistinctImageIDs()
	if len(got) != 2 || got[0] != "file_aaa" || got[1] != "file_bbb" {
		t.Fatalf("DistinctImageIDs = %v, want [file_aaa file_bbb]", got)
	}
}

// TestAltForID_DriftCase is the exact drift scenario: a generated image
// rendered through two <img> tags (file_aaa twice) followed by a second
// distinct image (file_bbb). DistinctImageIDs collapses this to
// [file_aaa, file_bbb], but Alts stays parallel to the raw, undeduplicated
// ImageURLs ([Foo, Foo, Bar]). An index-based lookup (Alts[i] where i is the
// position in the deduplicated id list) gives file_bbb -> Alts[1] == "Foo",
// which is wrong: file_bbb's own alt is Alts[2] == "Bar".
func TestAltForID_DriftCase(t *testing.T) {
	s := PageState{
		ImageURLs: []string{
			"https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs",
			"https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs",
			"https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs",
		},
		Alts: []string{
			"Generated image: Foo",
			"Generated image: Foo",
			"Generated image: Bar",
		},
	}
	if got := s.AltForID("file_aaa"); got != "Generated image: Foo" {
		t.Fatalf("AltForID(file_aaa) = %q, want %q", got, "Generated image: Foo")
	}
	if got := s.AltForID("file_bbb"); got != "Generated image: Bar" {
		t.Fatalf("AltForID(file_bbb) = %q, want %q", got, "Generated image: Bar")
	}
}

func TestAltForID_UnknownIDReturnsEmpty(t *testing.T) {
	s := PageState{
		ImageURLs: []string{"https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs"},
		Alts:      []string{"Generated image: Foo"},
	}
	if got := s.AltForID("file_zzz"); got != "" {
		t.Fatalf("AltForID(unknown) = %q, want empty", got)
	}
}

func TestAltForID_AltsShorterThanImageURLsDoesNotPanic(t *testing.T) {
	s := PageState{
		ImageURLs: []string{
			"https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs",
			"https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs",
		},
		Alts: []string{"Generated image: Foo"}, // shorter than ImageURLs
	}
	if got := s.AltForID("file_aaa"); got != "Generated image: Foo" {
		t.Fatalf("AltForID(file_aaa) = %q, want %q", got, "Generated image: Foo")
	}
	if got := s.AltForID("file_bbb"); got != "" {
		t.Fatalf("AltForID(file_bbb) = %q, want empty (Alts too short)", got)
	}
}

// TestAltForID_NonGeneratedURLInterleaved guards against a non-generated tag
// (a favicon, which FileIDFromURL maps to "") shifting the pairing between
// generated ids and their alts.
func TestAltForID_NonGeneratedURLInterleaved(t *testing.T) {
	s := PageState{
		ImageURLs: []string{
			"https://chatgpt.com/cdn/assets/favicon-x.svg",
			"https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs",
			"https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs",
		},
		Alts: []string{
			"",
			"Generated image: Foo",
			"Generated image: Bar",
		},
	}
	if got := s.AltForID("file_aaa"); got != "Generated image: Foo" {
		t.Fatalf("AltForID(file_aaa) = %q, want %q", got, "Generated image: Foo")
	}
	if got := s.AltForID("file_bbb"); got != "Generated image: Bar" {
		t.Fatalf("AltForID(file_bbb) = %q, want %q", got, "Generated image: Bar")
	}
}

func TestDoneRequiresQuietAndEnoughImages(t *testing.T) {
	img := func(n int) []string {
		var out []string
		for i := 0; i < n; i++ {
			out = append(out, "https://chatgpt.com/backend-api/estuary/content?id=file_"+string(rune('a'+i))+"&p=fs")
		}
		return out
	}
	cases := []struct {
		name string
		s    PageState
		want int
		done bool
	}{
		{"still loading", PageState{Loading: true, ImageURLs: img(1)}, 1, false},
		{"still streaming", PageState{Streaming: true, ImageURLs: img(1)}, 1, false},
		{"quiet but no image", PageState{}, 1, false},
		{"quiet with image", PageState{ImageURLs: img(1)}, 1, true},
		{"set incomplete", PageState{ImageURLs: img(2)}, 3, false},
		{"set complete", PageState{ImageURLs: img(3)}, 3, true},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := Done(c.s, c.want); got != c.done {
				t.Fatalf("Done = %v, want %v", got, c.done)
			}
		})
	}
}

File: test/state.test.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');

const { parseState, distinctImageIds, altForId, done } = require('../src/state');

const oneImage = `{"loading":false,"streaming":false,
"imageURLs":["https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs&sig=x"],
"alts":["Generated image: Teal Mountain"]}`;

test('parseState', () => {
  const s = parseState(oneImage);
  assert.equal(s.loading, false);
  assert.equal(s.streaming, false);
  assert.equal(s.imageURLs.length, 1);
});

test('parseState throws on malformed input, never falls back silently', () => {
  assert.throws(() => parseState('{not valid json'));
});

test('distinctImageIds', () => {
  // The spike saw the same generated image rendered by three <img> tags.
  const s = {
    imageURLs: [
      'https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs',
      'https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs',
      'https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs',
      'https://chatgpt.com/cdn/assets/favicon-x.svg',
    ],
    alts: [],
  };
  const got = distinctImageIds(s);
  assert.deepEqual(got, ['file_aaa', 'file_bbb']);
});

// The exact drift scenario: a generated image rendered through two <img>
// tags (file_aaa twice) followed by a second distinct image (file_bbb).
// distinctImageIds collapses this to [file_aaa, file_bbb], but alts stays
// parallel to the raw, undeduplicated imageURLs ([Foo, Foo, Bar]). An
// index-based lookup (alts[i] where i is the position in the deduplicated
// id list) gives file_bbb -> alts[1] == "Foo", which is wrong: file_bbb's
// own alt is alts[2] == "Bar". Ported verbatim from
// TestAltForID_DriftCase in internal/compose/state_test.go; it must fail
// against an index-based implementation.
test('altForId walks imageURLs and alts together, keyed by id (drift case)', () => {
  const s = {
    imageURLs: [
      'https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs',
      'https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs',
      'https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs',
    ],
    alts: ['Generated image: Foo', 'Generated image: Foo', 'Generated image: Bar'],
  };
  assert.equal(altForId(s, 'file_aaa'), 'Generated image: Foo');
  assert.equal(altForId(s, 'file_bbb'), 'Generated image: Bar');
});

test('altForId returns empty for an unknown id', () => {
  const s = {
    imageURLs: ['https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs'],
    alts: ['Generated image: Foo'],
  };
  assert.equal(altForId(s, 'file_zzz'), '');
});

test('altForId does not throw when alts is shorter than imageURLs', () => {
  const s = {
    imageURLs: [
      'https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs',
      'https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs',
    ],
    alts: ['Generated image: Foo'], // shorter than imageURLs
  };
  assert.equal(altForId(s, 'file_aaa'), 'Generated image: Foo');
  assert.equal(altForId(s, 'file_bbb'), '');
});

// Guards against a non-generated tag (a favicon, which fileIdFromUrl maps
// to '') shifting the pairing between generated ids and their alts.
test('altForId is unaffected by a non-generated url interleaved', () => {
  const s = {
    imageURLs: [
      'https://chatgpt.com/cdn/assets/favicon-x.svg',
      'https://chatgpt.com/backend-api/estuary/content?id=file_aaa&p=fs',
      'https://chatgpt.com/backend-api/estuary/content?id=file_bbb&p=fs',
    ],
    alts: ['', 'Generated image: Foo', 'Generated image: Bar'],
  };
  assert.equal(altForId(s, 'file_aaa'), 'Generated image: Foo');
  assert.equal(altForId(s, 'file_bbb'), 'Generated image: Bar');
});

test('done requires quiet and enough distinct images', () => {
  const img = (n) => {
    const out = [];
    for (let i = 0; i < n; i++) {
      out.push(`https://chatgpt.com/backend-api/estuary/content?id=file_${String.fromCharCode(97 + i)}&p=fs`);
    }
    return out;
  };
  const cases = [
    { name: 'still loading', s: { loading: true, streaming: false, imageURLs: img(1), alts: [] }, want: 1, expect: false },
    { name: 'still streaming', s: { loading: false, streaming: true, imageURLs: img(1), alts: [] }, want: 1, expect: false },
    { name: 'quiet but no image', s: { loading: false, streaming: false, imageURLs: [], alts: [] }, want: 1, expect: false },
    { name: 'quiet with image', s: { loading: false, streaming: false, imageURLs: img(1), alts: [] }, want: 1, expect: true },
    { name: 'set incomplete', s: { loading: false, streaming: false, imageURLs: img(2), alts: [] }, want: 3, expect: false },
    { name: 'set complete', s: { loading: false, streaming: false, imageURLs: img(3), alts: [] }, want: 3, expect: true },
  ];
  for (const c of cases) {
    assert.equal(done(c.s, c.want), c.expect, c.name);
  }
});
